### PR TITLE
Make http-to-bucket cloud function clean hashes before sending to big query

### PR DESCRIPTION
**Pull request by JonathanHallam**
Uses recursion to remove empty hashes from all layers of a given array or hash. This is because BigQuery can't handle the empty hashes.

This is further outlined in this issue - https://github.com/alphagov/govuk-knowledge-graph-gcp/issues/837

### CODE REVIEW

File: docker/http-to-bucket/app.rb [hunk 1]
```diff
@@ -82,6 +82,8 @@ def upload_response(project_id, bucket_name, object_name, http_response)
 
     request.logger.info "Parameters are valid"
 
+    clean(request)
```
Comment (nacnudus) [inline docker/http-to-bucket/app.rb:85]: It's the response that needs to be cleaned, as part of `upload_response()`, probably between lines 50 and 51, e.g. `clean(body_json)`.

File: docker/http-to-bucket/app.rb [hunk 2]
```diff
@@ -131,3 +133,16 @@ def upload_response(project_id, bucket_name, object_name, http_response)
   # which will be JSON-encoded into a response.
   return JMESPath.search(jmespath, JSON.parse(http_response.body))
 end
+
+def clean(obj)
```
Comment (nacnudus) [inline docker/http-to-bucket/app.rb:137]: This is a neat implementation. A couple of nitpicks:

* Could the name be more descriptive? Another kind of cleaning occurs in lines 51-59. If you feel like refactoring those lines into their own function too, that'd be welcome, but it's also okay to leave that for another time.
* Please could this function be defined above `FunctionsFramework.http "http_to_bucket" do |request|`, alongside the other helper functions.

File: docker/http-to-bucket/spec/app_spec.rb [hunk 3]
```diff
@@ -316,4 +316,72 @@ def build_request_url(path = "/", params = {})
...
+          a: {bar: "baz"},
+          b: {foo: "bar"},
+          c: {foo: {bar: "baz"}}
+        }
+
+        expect(clean(object)).to eq(expectation)
+      end
+    end
+
+    context "when it's given an array of many multi layered hash" do
```
Comment (nacnudus) [inline docker/http-to-bucket/spec/app_spec.rb:356]: This test seems redundant, given that earlier tests prove the array handling and recursion. I might be wrong. Not a big deal.

File: docker/http-to-bucket/spec/app_spec.rb [hunk 4]
```diff
@@ -316,4 +316,72 @@ def build_request_url(path = "/", params = {})
...
+          a: {bar: "baz"},
+          b: {foo: "bar"},
+          c: {foo: {bar: "baz"}}
+          },
+        ]
+
+        expect(clean(object)).to eq(expectation)
+      end
+    end
+  end
```
Comment (nacnudus) [inline docker/http-to-bucket/spec/app_spec.rb:386]: Please could you add a test that it doesn't remove empty arrays or empty strings?

```
object = {foo: []}
expectation = {foo: []}
```
```
object = {foo: ""}
expectation = {foo: ""}
```

**Review by nacnudus — CHANGES_REQUESTED** [review]
Nice implementation, thanks! I've asked for one important change, and a couple of minor things.

File: docker/http-to-bucket/spec/app_spec.rb (hunk 4 above)
Comment (nacnudus) [reply docker/http-to-bucket/spec/app_spec.rb:386]: Another thought: it might be worth testing this:
```
object = {foo: {bar: {}}
expectation = {}
```

**Review by nacnudus — APPROVED** [review]

File: docker/http-to-bucket/spec/app_spec.rb (hunk 4 above)
Comment (aaronfowles) [reply docker/http-to-bucket/spec/app_spec.rb:404]: Resolving as I can see that Duncan's test case suggestions have been implemented and they're passing.
